Prototype props table for design system documentation

aduth opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Closes DS-206

## Proposed Changes

* Adds script as part of `@automattic/components` package build to compile documentation metadata for documented components
* Updates design system documentation to use said metadata to display a table of props

This is meant to be an early, prototype implementation, subject to design refinement.

For review, consider:

* Is this workflow sensible and scalable?
* Can it support how we intend to use these props in a published site deployment process?
* Are the assumptions around how we can parse components accurate?

Additional notes:

* I had attempted to reuse Storybook components directly, but they are not especially well-suited to being used outside a Storybook context. Additionally, I don't know how much of the Storybook "experience" we would want to set as an expectation within our design system site. The established tooling around props parsing should be straight-forward enough to provide us the information we need.
* This uses a similar approach to [component documentation generation in WordPress](https://github.com/WordPress/gutenberg/blob/trunk/bin/api-docs/gen-components-docs/index.mjs)

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* To be used as part of the technical reference for a design system component

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Packages need to be built first:

```
yarn build-packages
```

Then, start the documentation site:

```
yarn workspace @automattic/design-system-docs start
```

Navigate to "Components" and verify that a props table is shown for [Badge](http://localhost:42987/components/ds/feedback/badge).

![image](https://github.com/user-attachments/assets/433f73b1-4caf-41bb-b2c7-aa66c5dcb35b)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Prototype props table for design system documentation

aduth commented on packages/components/bin/build-metadata.mts:
```diff
@@ -0,0 +1,76 @@
+
+const files = DOCUMENTED_COMPONENTS.map( ( component ) => `${ component }/index.tsx` );
+
+const pick = < O extends Record< string, any >, K extends Array< keyof O > >(
+	obj: O,
+	keys: K
+): Pick< O, K[ number ] > =>
+	Object.fromEntries( Object.entries( obj ).filter( ( [ key ] ) => keys.includes( key ) ) ) as Pick<
+		O,
+		K[ number ]
+	>;
+
+const omit = < O extends Record< string, any >, K extends Array< keyof O > >(
+	obj: O,
+	keys: K
+): Omit< O, K[ number ] > =>
+	Object.fromEntries(
+		Object.entries( obj ).filter( ( [ key ] ) => ! keys.includes( key ) )
+	) as Omit< O, K[ number ] >;
+
+const mapValues = < V, MV >(
+	obj: Record< string, V >,
+	fn: ( value: V ) => MV
+): Record< string, MV > =>
+	Object.fromEntries( Object.entries( obj ).map( ( [ key, value ] ) => [ key, fn( value ) ] ) );
```
This is a good example of where rolling our own utilities is not so fun 🙃 (Re: https://github.com/Automattic/wp-calypso/pull/103048#discussion_r2071495119)

mirka replied:
Who needs lodash when you can write your own lodash

mirka replied:
Jokes aside, I don't even remember lodash functions anymore so it may not necessarily be more readable for certain audiences 🙈

aduth commented on packages/components/bin/build-metadata.mts:
```diff
@@ -0,0 +1,76 @@
+	>;
+};
+
+const DOCUMENTED_COMPONENTS = [ 'breadcrumbs', 'summary-button' ];
+
+const files = DOCUMENTED_COMPONENTS.map( ( component ) => `${ component }/index.tsx` );
```
I think in the future this would ideally be something we don't maintain as a list and instead `glob( 'src/*/index.tsx' )`. But `react-docgen-typescript` is pretty slow at parsing, and we have a lot of components -- most of which we don't plan on keeping here long-term. So for now I went with an allowlist.

mirka replied:
The reason I ended up needing explicit [manifest files](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/schemas/docs-manifest.json) is to deal with subcomponents. The `Tabs` component that Marco is working on porting over (#103515) is a good example of the [complexity](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/tabs/docs-manifest.json) involved. I believe Storybook itself is getting it from the [`subcomponents` metadata](https://github.com/WordPress/gutenberg/blob/77266d7d3103227d149233c121a59bd4974c91d8/packages/components/src/tabs/stories/index.story.tsx#L26-L35).

Not saying we absolutely need JSON files for it, in fact there may be a less clunky way for us in the new UI package because we don't have to deal with a mixture of legacy components that expose their subcomponents in a different style, e.g. [`Card`](https://github.com/WordPress/gutenberg/blob/77266d7d3103227d149233c121a59bd4974c91d8/packages/components/src/index.ts#L52-L59) that doesn't use dot notation.

I also just realized that having a clear and canonical way to declare what our intended exports (component structures) are may help with tree shakeable CSS builds like you [prototyped](https://github.com/Automattic/wp-calypso/compare/add/components-css-modules#diff-93917c9c179f4abaa5afbd97fd3417257190a58706a86f5661b6561ad071d1d7).

aduth replied:
That's a good call for a more complex example. I incorporated a demonstration of support for subcomponents in d0d9da4401f . I don't know that we'll want to have a mapping like this long-term vs. something like a manifest file, and am open to feedback. I think this makes for a consolidated source of truth, though may not be especially scalable as we add more.

aduth commented on packages/components/package.json:
```diff
@@ -16,10 +16,15 @@
 			"calypso:src": "./styles/*"
+		},
+		"./metadata": {
+			"dsdocs:src": "./dist/metadata.js",
```
This and the inline Docusaurus plugin are meant to kinda "hide" this from the public interface of the package. In the future, if we want to standardize and expose this, we should probably just have this available as `"import"`

aduth replied:
I reverted this in 7f2d6ab3bf3 to use plain, "public" `import`. Docusaurus doesn't seem to be very compatible with using `conditionNames` overrides in the Webpack configuration. I was getting a bunch of errors in how the internal theme's CSS modules were being processed:

Example:

```
Syntax error: /Code/wp-calypso/node_modules/@docusaurus/theme-classic/lib/theme/AnnouncementBar/CloseButton/styles.module.css Unknown word "closeButton_rfix" (2:31)

  1 | // extracted by mini-css-extract-plugin
> 2 | export default {"closeButton":"closeButton_rfix"};
    |                               ^
```

aduth commented on apps/design-system-docs/docs/components/0-index.mdx:
Docusaurus sidebars are [autogenerated](https://docusaurus.io/docs/sidebar/autogenerated) as alphabetical, with the first subpage being the target of the top-level navigation. We can create our own explicit sidebar configuration, but this convention of numeric prefixes was [one suggested in the documentation](https://docusaurus.io/docs/sidebar/autogenerated#using-number-prefixes) and seems like an easier way to manage the single index page for each section.

aduth replied:
The numeric prefix broke the `/components/` path using the index file, despite [the documentation claiming it'd be removed from the URL paths](https://docusaurus.io/docs/sidebar/autogenerated#using-number-prefixes:~:text=By%20default%2C%20Docusaurus%20will%20remove%20the%20number%20prefix%20from%20the%20doc%20id%2C%20title%2C%20label%2C%20and%20URL%20paths.).

Instead, I reverted this back to `index.mdx` and used frontmatter to assign the sidebar position. Hopefully we should only have to do this for the `index.mdx` file and not individual components.

aduth commented on apps/design-system-docs/src/components/props-table.tsx:
```diff
@@ -0,0 +1,53 @@
+import metadata from '@automattic/components/metadata';
+import { VisuallyHidden } from '@wordpress/components';
```
Problem: I'm running into some strange errors when building the site, specifically during static site generation. I'm wondering if there might be some incompatibility with using WordPress components in a statically rendered context?

I don't see Gutenberg rendering this Emotion Cache provider anywhere. I also _tried_ to ["swizzle"](https://docusaurus.io/docs/swizzling) a wrapper to provide the context, but the error persisted.

Possible solution could be to use `ScreenReaderText` from `@automattic/components`, or create a duplicate copy in this project, or avoid the visually-hidden content altogether.

```
      [errors]: [
        Error: Can't render static file for pathname "/components/summary-button"
            at generateStaticFile (/Code/wp-calypso/node_modules/@docusaurus/core/lib/ssg/ssg.js:167:15)
            at processTicksAndRejections (node:internal/process/task_queues:105:5)
            at runNextTicks (node:internal/process/task_queues:69:3)
            at process.processImmediate (node:internal/timers:459:9)
            at async /Code/wp-calypso/node_modules/p-map/index.js:57:22 {
          [cause]: Error: The `useCx` hook should be only used within a valid Emotion Cache Context
              at server.bundle.js:7113:13
              at useContextSystem (server.bundle.js:7176:19)
              at UnconnectedVisuallyHidden (server.bundle.js:7608:7)
              at Uc (server.bundle.js:4062:44)
              at Xc (server.bundle.js:4067:395)
              at Z (server.bundle.js:4070:89)
              at Yc (server.bundle.js:4073:98)
              at $c (server.bundle.js:4072:140)
              at Z (server.bundle.js:4070:345)
              at Xc (server.bundle.js:4065:476)
```

aduth replied:
I opted to replace in fcc25223e68 with static text. This table will need some design refinement anyways, so let's make it a future problem to solve.

aduth committed:
Expose metadata as plain import

aduth committed:
Use static "Required" text

mirka commented on packages/components/bin/build-metadata.mts:
```diff
@@ -0,0 +1,76 @@
+	return [
+		displayName,
+		{
+			props: mapValues( omit( props, [ 'ref', 'key' ] ), ( value ) =>
```
We should [match](https://github.com/WordPress/gutenberg/blob/77266d7d3103227d149233c121a59bd4974c91d8/bin/api-docs/gen-components-docs/index.mjs#L18-L25) the `react-docgen-typescript` config that Storybook uses. This will prevent unhelpfully long prop lists like we can see in the `Breadcrumbs` component.

Related but separate problem (probably out of scope for this PR): How/where to document "rest props" (https://github.com/WordPress/gutenberg/issues/67979)

aduth replied:
I updated to use the same options in 6ec226d0eb5, which helped to remove the need for explicit omission of `key` and `ref`. I do notice that it sometimes excludes some props that would be valuable to include, like `Tabs.TabList` `children`. I think it might be excluding them because we reference the `children` type from AriaKit ?

https://github.com/Automattic/wp-calypso/blob/47dd847a9587287df79e6e6036ab37d6f06d4965/packages/components/src/tabs/types.ts#L94-L100

ciampo replied:
Likely — I believe we currently document only first-party types ?

mirka commented on packages/components/bin/build-metadata.mts:
```diff
@@ -0,0 +1,76 @@
+await mkdir( 'dist/types', { recursive: true } );
+
+await Promise.all( [
+	writeFile( 'dist/metadata.js', `export default ${ JSON.stringify( metadata, null, 2 ) };` ),
```
I'm thinking what the workflow would be like on PRs that make docs-affecting changes to the components package. Ideally, we'd prompt two checks as part of the code review:

1. Did prop types change in the way you expected? (When dealing with types composed from other types using `Pick`/`Omit`, you sometimes fail to notice that your type changes affected the prop types of other components in ways you didn't intend.)
2. Are your TSDocs rendering correctly on the site? (Badly formatted markdown, unsupported `@tags`, etc)

What do you think? Do a diff of `dist/metadata.js` on CI, and post a comment with live links to the doc site?

aduth replied:
I think this ties well into my comment at https://github.com/Automattic/wp-calypso/pull/103778#issuecomment-2916691963, specifically about path changes where we might expect to need a preview link. I think we could likely build upon the GitHub Action in #103778 to do more consideration around changes to the manifest, and direct links to affected components. 

I'd probably reserve that for future enhancement though, once both this pull request and #103778 have been merged. I can create a tracking issue in Linear for that if it helps.

aduth committed:
Use consistent options from Gutenberg parsing to exclude irrelevant props

See:

- https://github.com/Automattic/wp-calypso/pull/103561#discussion_r2103080017
- https://github.com/WordPress/gutenberg/blob/77266d7d3103227d149233c121a59bd4974c91d8/bin/api-docs/gen-components-docs/index.mjs#L18-L25

aduth committed:
Fix defaultValue type as nullable

aduth committed:
Add comments explaining purpose of utility functions

aduth committed:
Add more inline code comments

aduth committed:
Use Docusaurus Webpack aliases instead of subpath imports

Consistency across documentation, and with expectations for using Docusaurus

aduth committed:
Move PropsTable into folder structure

Consistency with other components LinkCard, LinkCardList

aduth committed:
yarn dedupe

aduth committed:
Incorporate support for nested components
